src/core_functions/scalar/list/list_reduce.cpp

```diff
@@ -0,0 +1,217 @@
+			input_chunk.data[slice_offset + 2 + i].Reference(slices.back());
+		}
+	}
+
+	result_chunk.Reset();
+	result_chunk.SetCardinality(new_count);
+
+	execute_info.expr_executor->Execute(input_chunk, result_chunk);
+
+	// use the result chunk to update the left slice
+	execute_info.left_slice.Reference(result_chunk.data[0]);
+}
+
+void LambdaFunctions::ListReduceFunction(duckdb::DataChunk &args, duckdb::ExpressionState &state,
+                                         duckdb::Vector &result) {
+	// Initializes the left slice from the list entries, active rows, the expression executor and the input types
+	bool completed = false;
+	LambdaFunctions::LambdaInfo info(args, state, result, completed);
+	if (completed) {
+		return;
+	}
+
+	ReduceExecuteInfo execute_info(info, state.GetContext());
+
+	// Since the left slice references the result chunk, we need to create two result chunks.
+	// This means there is always an empty result chunk for the next iteration,
+	// without the referenced chunk having to be reset until the current iteration is complete.
+	DataChunk odd_result_chunk;
+	odd_result_chunk.Initialize(Allocator::DefaultAllocator(), {info.lambda_expr->return_type});
+
+	DataChunk even_result_chunk;
+	even_result_chunk.Initialize(Allocator::DefaultAllocator(), {info.lambda_expr->return_type});
+
+	idx_t loops = 0;
+	// Execute reduce until all rows are finished
+	while (!execute_info.active_rows.empty()) {
+		if (loops % 2) {
+			ExecuteReduce(loops, execute_info, info, odd_result_chunk);
+			even_result_chunk.Reset();
+		} else {
+			ExecuteReduce(loops, execute_info, info, even_result_chunk);
+			odd_result_chunk.Reset();
+		}
+
+		loops++;
+	}
+
+	if (info.is_all_constant && !info.has_side_effects) {
+		info.result.SetVectorType(VectorType::CONSTANT_VECTOR);
+	}
+}
+
+static unique_ptr<FunctionData> ListReduceBind(ClientContext &context, ScalarFunction &bound_function,
+                                               vector<unique_ptr<Expression>> &arguments) {
+
+	// the list column and the bound lambda expression
+	D_ASSERT(arguments.size() == 2);
+	if (arguments[1]->expression_class != ExpressionClass::BOUND_LAMBDA) {
+		throw BinderException("Invalid lambda expression!");
+	}
+
+	arguments[0] = BoundCastExpression::AddArrayCastToList(context, std::move(arguments[0]));
+
+	auto &bound_lambda_expr = arguments[1]->Cast<BoundLambdaExpression>();
+	if (bound_lambda_expr.parameter_count < 2 || bound_lambda_expr.parameter_count > 3) {
+		throw BinderException("list_reduce expects a function with 2 or 3 arguments");
+	}
+	auto has_index = bound_lambda_expr.parameter_count == 3;
+
+	unique_ptr<FunctionData> bind_data = LambdaFunctions::ListLambdaPrepareBind(arguments, context, bound_function);
+	if (bind_data) {
+		return bind_data;
+	}
+
+	auto list_child_type = arguments[0]->return_type;
+	list_child_type = ListType::GetChildType(list_child_type);
+
+	auto cast_lambda_expr =
```

Comment:
From what I remember about looking into this, it is problematic, especially when binding the lambda function. We bind the lambda function with the two input parameter types, `list_child_type`. After executing the lambda expression on the first two input elements (`list_child_type`), the result value **becomes** the next input. Now, if we cannot cast the result value to the input type, `list_child_type`, the expression executor does not know what input to expect for the lambda parameter: the list child type or the lambda expression return type.

Another way this might work is binding the lambda function once with the `list_child_type`, testing if we can cast the child entries to the return type, and then binding **again** with the return type. In this scenario, the input types have to be castable to the lambda function return type. In the first (current) scenario, the lambda return type has to be castable to the input parameter type.

Reply:
Yes that makes sense - scratch that, let's leave it like it is